dom.go

```diff
@@ -252,19 +268,59 @@ func (h *HTML) reconcile(prev *HTML) []Mounter {
 		h.node.Set("innerHTML", h.innerHTML)
 	}
 
-	return h.reconcileChildren(prev, nil)
+	return h.reconcileChildren(prev)
 }
 
 // reconcileChildren reconciles children of the current HTML against a previous
 // render's DOM nodes.
-func (h *HTML) reconcileChildren(prev *HTML, insertBefore *HTML) (pendingMounts []Mounter) {
-	// TODO better list element reuse
+func (h *HTML) reconcileChildren(prev *HTML) (pendingMounts []Mounter) {
+	var keyed bool
 	for i, nextChild := range h.children {
-		// TODO(pdf): Add tests for node equality
-		new := h.node != prev.node
-		if i >= len(prev.children) || new {
-			if nextChildList, ok := nextChild.(List); ok {
-				nextChildList.reconcile(h.node, insertBefore, nil)
+		// Massage concrete type if necessary.
+		switch v := nextChild.(type) {
+		case *HTML:
+			if v == nil {
+				nextChild = nil
+				h.children[i] = nextChild
+			}
+		case List:
+			nextChild = childList{html: &HTML{children: v}}
+			h.children[i] = nextChild
+		}
+
+		// TODO(pdf): Add tests for node equality, keyed children
+		var (
+			new                   = h.node != prev.node
+			nextKey               string
+			stableKey             bool
+			insertBeforeKeyedNode jsObject
+		)
+		if keyer, ok := nextChild.(Keyer); ok {
+			nextKey = keyer.Key()
+			if len(h.keyMap) > 0 && nextKey == "" {
+				panic("vecty: all siblings must have keys when using keyed elements")
+			}
+			if nextKey != "" {
+				if h.keyMap == nil {
+					h.keyMap = make(map[string]ComponentOrHTML)
+				}
+				if _, exists := h.keyMap[nextKey]; exists {
+					panic("vecty: duplicate sibling key: " + nextKey)
+				}
+				h.keyMap[nextKey] = nextChild
+			}
+		}
+		if !keyed {
+			keyed = len(h.keyMap) > 0
+		}
+
+		if (i >= len(prev.children) && !keyed) || new {
```

Comment:
1. This block is purely responsible for handling creation of new child elements that the previous DOM node's children did not have, so we should document that now.

2. Why does this condition on `!keyed` ? Why would new children that are keyed _not_ be handled here? I don't understand.

Reply:
1. Ack
2. Because position can not be used for anything when dealing with keyed children - the feature is designed to deal with children that may move position, and that means they may appear at a higher index on the current render, than the last render had.  Once we know we're dealing with keyed children, the key is the only authoritative index.